cpp/src/arrow/testing/process.cc

```diff
@@ -0,0 +1,218 @@
+#else
+    ARROW_UNUSED(ec);
+    return Status::NotImplemented("Not available on this system");
+#endif
+    if (ec) {
+      // XXX fold this into the Status class?
+      return Status::IOError("Can't resolve current exe: ", ec.message());
+    } else {
+      return Status::OK();
+    }
+  }
+
+ public:
+  Impl() {
+    // Get a copy of the current environment.
+    for (const auto& kv : process::environment::current()) {
+      env_[kv.key()] = kv.value().string();
+    }
+  }
+
+  ~Impl() { process_ = nullptr; }
+
+  Status SetExecutable(const std::string& name) {
+    executable_ = process::environment::find_executable(name);
+    if (executable_.empty()) {
+      // Search the current executable directory as fallback.
+      boost::filesystem::path current_exe;
+      auto status = ResolveCurrentExecutable(&current_exe);
+      if (status.ok()) {
+        std::unordered_map<process::environment::key, process::environment::value> env = {
+            {"PATH", current_exe.parent_path().string()},
+        };
+        executable_ = process::environment::find_executable(name, env);
+      }
+    }
+    if (executable_.empty()) {
+      return Status::IOError("Failed to find '", name, "' in PATH");
+    }
+    return Status::OK();
+  }
+
+  Status SetArgs(const std::vector<std::string>& args) {
+    args_ = args;
+    return Status::OK();
+  }
+
+  Status SetEnv(const std::string& name, const std::string& value) {
+    // Workaround for https://github.com/boostorg/process/issues/365
+    env_[name] = std::string(value);
+    return Status::OK();
+  }
+
+  Status SetReadyErrorMessage(const std::string& marker) {
+    marker_ = marker;
+    return Status::OK();
+  }
+
+  Status Execute() {
+    try {
+      process::process_environment env(env_);
+      if (marker_.empty()) {
+        // We can't use std::make_unique<process::process>.
+        process_ = std::unique_ptr<process::process>(
+            new process::process(ctx_, executable_, args_, env));
+        return Status::OK();
+      }
+
+      asio::readable_pipe stderr(ctx_);
+      // We can't use std::make_unique<process::process>.
+      process_ = std::unique_ptr<process::process>(new process::process(
+          ctx_, executable_, args_, env, process::process_stdio{{}, {}, stderr}));
+      std::stringstream buffered_output;
+      std::array<char, 1024> buffer;
+      std::string line;
+      auto timeout = std::chrono::seconds(10);
+      std::chrono::time_point<std::chrono::steady_clock> end =
+          std::chrono::steady_clock::now() + timeout;
+      while (process_->running() && std::chrono::steady_clock::now() < end) {
+        auto read_bytes = stderr.read_some(asio::buffer(buffer.data(), buffer.size()));
+        if (buffered_output.eof()) {
+          buffered_output.clear();
+          auto last = buffered_output.str().size();
+          buffered_output.seekg(last);
+          buffered_output.seekp(last);
+        }
```

Comment:
What does this do? Can you add an explanatory comment?

Reply:
It's for clearing a EOS bit. I'll add a comment.